libuavcan/include/uavcan/driver/can.hpp

```diff
@@ -18,6 +19,32 @@ namespace uavcan
  */
 enum { MaxCanIfaces = 3 };
 
+/**
+ * Valid message data lengths.
+ *
+ * CAN DLCs are only 4-bits long so FD had to use 9-15 to encode
+ * various lengths up to 64 bits.
+ */
+enum class MessageDLC : uint_fast8_t {
```

Comment:
Better type name: `CANFrameDLC` (otherwise there'd be a conflict of terms: in UAVCAN, "message" is a special case of data transfer alongside service request and service response).

You are using scoped enums here, which is great (if we really decided to leave C++03 behind, that is). In this case the prefix `MessageDLC_` becomes redundant: `MessageDLC::MessageDLC_0` - no way.

Actually, come to think of it, do we really need this enum? You have defined two mappings elsewhere which map between data size and DLC and back. Can we not just use these? Or, better yet, there should be constexpr functions for that, for extra compile-time and run-time safety:

```c++
constexpr std::uint8_t lengthToDLC(std::uint8_t length)
{
    UAVCAN_ASSERT(length <= 64);   // Legal since C++14
    return ...; // you get the point
}
```

Another reason this is better than an enum is because enums are not supposed to be casted to integral values. This disagrees with the spirit of MISRA. I understand that there are a lot of placed in libuavcan where I did just that, but that was because I had no other option at the time of writing the original codebase - I had to keep it compatible with C++03 (ugh).

Reply:
Well, it's a lot safer to cast to integral values in C++11 given that you can specify and safely query the underlying type. That said, I'll give your comments some consideration and see what I can come up with.

Reply:
I only find 4-5-2 in MISRA c++ that may have something to say about enumeration values but this prohibits using them with operators other than `[]` which isn't what we are doing with the CanFrameDLC (renamed) type. Unless I'm missing some other rule I don't think my use of the enumeration's values violates any MISRA requirements or advisories.

I do think it's good to use this enumeration since it ensures the compiler catches places where the programmer mixes up the DLC and the data length. Making the two distinct types (i.e. DLC is CanFrameDLC and length is an integer type) give the compiler stricter semantics to work with.

Reply:
So how about:
```
enum class CanFrameDLC : uint_fast8_t {
    code_for_length_0 = 0,       /**< Data Length Code: 0 bytes */
    code_for_length_1 = 1,       /**< Data Length Code: 1 bytes */
   ...
    code_for_length_12 = 9,   /**< Data Length Code: 12 bytes */
    ...
    code_for_length_64 = 15,  /**< Data Length Code: 64 bytes */
    invalid_code = 16 /**< Not a valid DLC. */
};
```

Reply:
Okay. How about making it camel case though?

Reply:
sure but which of these do you want:
1. `codeForLength_n`
3. `codeForLengthN`

Reply:
`CodeForLengthX`? Or is it PascalCase. I think it's PascalCase. Sorry for the confusion above.